---
title: "Indirect Selection"
author: "Jinliang Yang"
date: "02-24-2022"
output: pdf_document
---

## Path Normalization

````{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('..//')) 
#library(tidyverse)
```

--------------------------------

```{r}
geno <- read.table("data/geno.txt", header=FALSE)
dim(geno)
head(geno)
names(geno) <- c("chr", "pos", "ref", "alt", paste0("plant", 1:20))
```


### Create an F2 population

- Parent 1: plant1
- Parent 2: plant18

```{r}
### Just sample 100 markers from this Mt chr
set.seed(12579)
markers <- sample(1:nrow(geno), size=100)
f <- geno[sort(markers), c("chr", "pos", "ref", "alt", "plant1", "plant18")]
# select just one haplotype
f$plant1 <- gsub("/.*", "", f$plant1)
f$plant18 <- gsub("/.*", "", f$plant18)
# recoding to use -1,0,1
f[f==0] <- -1
# simulate the recombination rate
f$cM <- f$pos/5000
```


### G

```{r}
install.packages("devtools")
library(devtools)
install_github("lian0090/simuPoisson")
```

```{r}
library(simuPoisson)
data(parentsGeno)
data(map)
#simulate 10 individuals
progeny=simuPoisson(parentsGeno,map$chr,map$cM,10)
```


We simulated an F2 population with 200 individuals and 100 SNP markers

```{r}
pgeno <- t(f[, c("plant1", "plant18")])
pgeno <- apply(pgeno, 2, as.numeric)
f2 <- simuPoisson(pgeno, f$chr, f$cM, 200)

f2 <- as.data.frame(f2)
names(f2) <- paste0(f$chr, "_", f$pos)
write.table(f2, "data/f2_geno.csv", sep=",", quote=FALSE)
```

---

# Genotype frq and allele frq

Coding `-1, 0, 1`

Note, not all markers are segregting.
```{r}
table(f2[,6])
```

### Observed allele frequency

```{r}
# For A1 allele
p <- (52*2+92)/((52+92+56)*2)
# For A2 allele
q <- (56*2+92)/((52+92+56)*2)
```


### Observed genotype frequency

```{r}
# For A1A1 genotype
A1A1 <- 52/(52+92+56)
# For A1A2 genotype
A1A2 <- 92/(52+92+56)
# For A2A2 genotype
A2A2 <- 56/(52+92+56)
```


### Predicted genotype frequency

```{r}
p^2
2*p*q
q^2
```
```{r}
chisq.test(rbind(c(A1A1, A1A2, A2A2), c(p^2, 2*p*q, q^2)))
```

---

```{r}
p1 <- rnorm(52, mean=60, sd=10)
p2 <- rnorm(92, mean=90, sd=11)
p3 <- rnorm(56, mean=80, sd=12)

pheno <- f2[, 1:6]
pheno[pheno[, 6]==-1, 5] <- p1
pheno[pheno[, 6]==0, 5] <- p2
pheno[pheno[, 6]==1, 5] <- p3

pheno <- pheno[, 5:6]
names(pheno)[1] <- c("height", "plantid")
pheno$plantid <- paste0("plant", 1:nrow(pheno))
write.table(pheno[, c(1,3)], "data/f2_pheno.csv", sep=",", row.names = FALSE, quote=FALSE)
```


```{r}
pheno <- read.csv("data/f2_pheno.csv")

hist(pheno$height, main="Plant Height", xlab="Value (inch)", breaks=20)
```


```{r}
library(ggplot2)

# Change line colors by groups
ggplot(pheno, aes(x=height)) +
  #facet_wrap(vars(trait), scales = "free") +
  geom_histogram(aes(y=..density..), position="identity", alpha=0.8, bins=20)+
  geom_density(alpha=0.38)+
  #scale_color_manual(values=c("#E69F00", "#56B4E9", "#fe6f5e"))+
  scale_color_manual(values=c("#E69F00"))+
  #scale_fill_manual(values=c("#E69F00", "#56B4E9", "#fe6f5e"), labels=c("HE", "HO", "Nip"))+
  labs(title="", x="Plant Height", y = "Density")+
  #guides(color=FALSE, fill=guide_legend(title="")) +
  theme_classic() +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.y = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          )

```


```{r}
gp <- cbind(pheno, f2)
```


Let's find out $a$ and $d$ at Marker `Mt_24242`

```{r}
u <- mean(gp$height) # population mean
# A1A1
h1 <- mean(subset(gp, Mt_24242 == -1)$height) 
# A1A2
h12 <- mean(subset(gp, Mt_24242 == 0)$height)
# A2A2
h2 <- mean(subset(gp, Mt_24242 == 1)$height)
```

```{r}

ggplot(gp, aes(x=as.factor(Mt_24242), y=height, color=as.factor(Mt_24242))) +
    geom_boxplot() +
    geom_jitter(color="black", size=1, alpha=0.9) +
    scale_color_manual(values=c("#E69F00", "#56B4E9", "#fe6f5e"))+
    labs(title="Mt_24242", y="Plant Height", x = "Genotype")+
    theme_classic() +
    guides(color=FALSE) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.y = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          )
```


```{r}
a <- (h2 - h1)/2
midpoint <- h1+a
d <- h12 - midpoint
```

---

# Allele Substitution Effect

The average effect of A1 and A2:

```{r}
alpha <- a + d*(q - p)
alpha1 <- q*alpha
alpha2 <- -p*alpha
```

# Genotypic values

The Breeding value associated with 

```{r}
bv2 = u+alpha2 + alpha2
bv1 = u+alpha1 + alpha1
bv12 = u+alpha1 + alpha2
```



```{r}
plot(c(0, 1, 2), c(h1, h12, h2), xlab="Genotype",ylab="", cex.lab=1.5, xaxt="n", pch=17, cex=2, col="red", ylim=c(60, 95)); 
axis(1, at=c(0, 1, 2), labels=c("A1A1", "A1A2", "A2A2")); 
mtext("Breeding Value", side = 4, line = 1, cex=1.5, col="blue"); 
mtext("Genotypic Value", side = 2, line = 2, cex=1.5, col="red")
points(c(0, 1, 2), c(bv2, bv12, bv1), cex=2, col="blue")
lines(c(0, 1, 2), c(bv2, bv12, bv1), lwd=2, col="blue")
```


# Additive and dominance variance

P = A + D + E

```{r}
Vp <- var(gp$height)
```

```{r}
Va <- 2*p*q*(a + d*(q - p))^2
Vd <- (2*p*q*d)^2
Vg <- Va + Vd
```

```{r}
h2 <- Va/Vp
H2 <- Vg/Vp
```


